sdk/identity/azure-identity/src/client_assertion_credential.cpp

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "azure/identity/client_assertion_credential.hpp"
+
+#include "private/identity_log.hpp"
+#include "private/package_version.hpp"
+#include "private/tenant_id_resolver.hpp"
+#include "private/token_credential_impl.hpp"
+
+#include <azure/core/internal/json/json.hpp>
+
+using Azure::Identity::ClientAssertionCredential;
+using Azure::Identity::ClientAssertionCredentialOptions;
+
+using Azure::Core::Context;
+using Azure::Core::Url;
+using Azure::Core::_internal::StringExtensions;
+using Azure::Core::Credentials::AccessToken;
+using Azure::Core::Credentials::AuthenticationException;
+using Azure::Core::Credentials::TokenRequestContext;
+using Azure::Core::Http::HttpMethod;
+using Azure::Identity::_detail::IdentityLog;
+using Azure::Identity::_detail::TenantIdResolver;
+using Azure::Identity::_detail::TokenCredentialImpl;
+
+namespace {
+bool IsValidTenantId(std::string const& tenantId)
+{
+  const std::string allowedChars = ".-";
+  if (tenantId.empty())
+  {
+    return false;
+  }
+  for (auto const c : tenantId)
+  {
+    if (allowedChars.find(c) != std::string::npos)
+    {
+      continue;
+    }
+    if (!StringExtensions::IsAlphaNumeric(c))
+    {
+      return false;
+    }
+  }
+  return true;
+}
+} // namespace
+
+ClientAssertionCredential::ClientAssertionCredential(
+    std::string tenantId,
+    std::string clientId,
+    std::function<std::string(Context const&)> const& assertionCallback,
+    ClientAssertionCredentialOptions const& options)
+    : TokenCredential("ClientAssertionCredential"), m_assertionCallback(assertionCallback),
+      m_clientCredentialCore(tenantId, options.AuthorityHost, options.AdditionallyAllowedTenants)
+{
+  bool isTenantIdValid = IsValidTenantId(tenantId);
+  if (!isTenantIdValid)
+  {
+    IdentityLog::Write(
+        IdentityLog::Level::Warning,
+        "Invalid tenant ID provided  for " + GetCredentialName()
+            + ". The tenant ID must be a non-empty string containing only alphanumeric characters, "
+              "periods, or hyphens. You can locate your tenant ID by following the instructions "
+              "listed here: https://learn.microsoft.com/partner-center/find-ids-and-domain-names");
+  }
+  if (clientId.empty())
+  {
+    IdentityLog::Write(
+        IdentityLog::Level::Warning, "No client ID specified for " + GetCredentialName() + ".");
+  }
+  if (!assertionCallback)
+  {
+    IdentityLog::Write(
+        IdentityLog::Level::Warning,
+        "The assertionCallback must be a valid function that returns assertions for "
```

Comment:
I think we can be less general, i.e. not "valid function that returns", but "not null".

Reply:
I am not sure I see the value in being that specific. An "empty/no target" `std::function` would still fail. The invalid case isn't only if null is passed.